OpenRA.Game/Primitives/LongBitSet.cs

```diff
@@ -0,0 +1,175 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2019 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace OpenRA.Primitives
+{
+	static class LongBitSetAllocator<T> where T : class
+	{
+		static readonly Cache<string, long> Bits = new Cache<string, long>(Allocate);
+		static long nextBits = 1;
+
+		static long Allocate(string value)
+		{
+			lock (Bits)
+			{
+				var bits = nextBits;
+				nextBits <<= 1;
+				return bits;
+			}
+		}
+
+		public static long GetBits(string[] values)
+		{
+			long bits = 0;
+			lock (Bits)
+				foreach (var value in values)
+					bits |= Bits[value];
+
+			return bits;
+		}
+
+		public static long GetBitsNoAlloc(string[] values)
+		{
+			// Map strings to existing bits; do not allocate missing values new bits
+			long bits = 0;
+
+			lock (Bits)
+			{
+				foreach (var value in values)
+				{
+					long valueBit;
+					if (Bits.TryGetValue(value, out valueBit))
+						bits |= valueBit;
+				}
+			}
+
+			return bits;
+		}
+
+		public static IEnumerable<string> GetStrings(long bits)
+		{
+			var values = new List<string>();
+			lock (Bits)
+				foreach (var kvp in Bits)
+					if ((kvp.Value & bits) != 0)
+						values.Add(kvp.Key);
+
+			return values;
+		}
+
+		public static bool BitsContainString(long bits, string value)
+		{
+			long valueBit;
+			lock (Bits)
+				if (!Bits.TryGetValue(value, out valueBit))
+					return false;
+			return (bits & valueBit) != 0;
+		}
+	}
+
+	public struct LongBitSet<T> : IEnumerable<string>, IEquatable<LongBitSet<T>> where T : class
```

Comment:
Can you add a description/summary here explaining that this is an optimized version of `BitSet` to be used when there are guaranteed to be no more than 64 values?